dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-fee-record-payment-entity-groups-to-keying-sheet.ts

```diff
@@ -1,65 +1,88 @@
 import Big from 'big.js';
-import { FeeRecordEntity, FeeRecordStatus, KeyingSheetAdjustment } from '@ukef/dtfs2-common';
+import { orderBy } from 'lodash';
+import { FeeRecordEntity, FeeRecordStatus } from '@ukef/dtfs2-common';
 import { FeeRecordPaymentEntityGroup } from '../../../../../helpers';
-import { KeyingSheet, KeyingSheetRow } from '../../../../../types/fee-records';
-import { mapFeeRecordEntityToKeyingSheetRowStatus, mapFeeRecordEntityToReportedPayments } from '../../../../../mapping/fee-record-mapper';
+import { KeyingSheet, KeyingSheetFeePayment, KeyingSheetRow } from '../../../../../types/fee-records';
+import { mapFeeRecordEntityToReportedPayments } from '../../../../../mapping/fee-record-mapper';
+import { mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments, mapPaymentEntityToKeyingSheetFeePayment } from '../../../../../mapping/keying-sheet-mapping';
 
-const getKeyingSheetAdjustmentForAmount = (amount: number | null): KeyingSheetAdjustment | null => {
-  if (amount === null) {
-    return null;
-  }
+const STATUSES_OF_FEE_RECORDS_TO_DISPLAY_ON_KEYING_SHEET: FeeRecordStatus[] = ['READY_TO_KEY', 'RECONCILED'];
 
-  const amountAbsoluteValue = new Big(amount).abs().toNumber();
-  if (amountAbsoluteValue === 0) {
-    return {
-      amount: 0,
-      change: 'NONE',
-    };
-  }
+/**
+ * Creates keying sheet rows for fee record payment groups
+ * with many fee records and many payments by greedily
+ * splitting payment amounts across the fee record fees
+ * sorted in descending order
+ * @param param - The fee record payment entity groups
+ * @param param.feeRecords - The fee record entities
+ * @param param.payments - The payment entities
+ * @returns The keying sheet
+ */
+const getKeyingSheetRowsWithPaymentsAcrossFeeRecords = ({ feeRecords, payments }: FeeRecordPaymentEntityGroup): KeyingSheetRow[] => {
+  const feeRecordsSortedDescending = orderBy(feeRecords, [(feeRecord) => feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency()], ['desc']);
+  let allFeePaymentsSortedAscending = orderBy(payments, ['amount'], ['asc']).map(mapPaymentEntityToKeyingSheetFeePayment);
 
-  return {
-    amount: amountAbsoluteValue,
-    change: amount > 0 ? 'INCREASE' : 'DECREASE',
-  };
-};
+  return feeRecordsSortedDescending.map((feeRecord) => {
+    const feePaymentsForRow: KeyingSheetFeePayment[] = [];
+    let remainingFeeRecordAmount = new Big(feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency());
+    while (remainingFeeRecordAmount.gt(0)) {
```

Comment:
I have a feeling something will go wrong if there is a payment matching tolerance in effect and the fee record amount total is greater than the payments total. Either this will hang forever or if it doesn't it will be because something inside throws is my suspicion, I don't know off the top of my head what happens when you attempt to pop from an empty array

Reply:
Popping from an empty array returns `undefined` I believe - technically, `pop` on an array of type `T[]` returns `T | undefined`, but I am overriding the with `pop()!`. Not sure how we want to integrate this with tolerance, but it is definitely going to require some more thinking on my end. I'll probably address this change from a test-driven POV and see what I can come up with